cmd/remote_enforcer.go

```diff
@@ -0,0 +1,291 @@
+	l4FlowHash := tcpPacket.L4FlowHash()
+	payload := &enforcer.StatsPayload{ContextID: contextID,
+		Tags:   tags,
+		Action: action,
+		Mode:   mode,
+		Source: sourceID,
+		Packet: tcpPacket,
+	}
+
+	c.Lock()
+	c.Flowentries.PushBack(&collectorentry{L4FlowHash: l4FlowHash, entry: payload})
+	c.Unlock()
+}
+
+//CollectContainerEvent exported
+//This event should not be expected here in the enforcer process inside a particular container context
+func (c *CollectorImpl) CollectContainerEvent(contextID string, ip string, tags policy.TagsMap, event string) {
+}
+
+//Server exported
+type Server struct {
+	MutualAuth  bool
+	Validity    time.Duration
+	SecretType  tokens.SecretsType
+	ContextID   string
+	CAPEM       []byte
+	PublicPEM   []byte
+	PrivatePEM  []byte
+	rpcchannel  string
+	rpchdl      *rpcwrapper.RPCWrapper
+	StatsClient *rpcwrapper.RPCWrapper
+	Enforcer    enforcer.PolicyEnforcer
+	Collector   collector.EventCollector
+	Supervisor  supervisor.Supervisor
+}
+
+type StatsClient struct {
+	collector *CollectorImpl
+	server    *Server
+	FlowCache *cache.Cache
+	Rpchdl    *rpcwrapper.RPCWrapper
+}
+
+func (s *StatsClient) SendStats() {
+	//We are connected and lets pack and ship
+	rpcpayload := &rpcwrapper.StatsPayload{}
+	var request rpcwrapper.Request
+	var response rpcwrapper.Response
+	var statsInterval time.Duration
+	rpcpayload.NumFlows = 0
+	EnvstatsInterval, err := strconv.Atoi(os.Getenv("STATS_INTERVAL"))
+
+	if err == nil && EnvstatsInterval != 0 {
+		statsInterval = time.Duration(EnvstatsInterval) * time.Second
+	} else {
+		statsInterval = defaulttimeInterval * time.Second
+	}
+
+	starttime := time.Now()
+	for {
+		s.collector.Lock()
+		for !(s.collector.Flowentries.Len() > 0) {
+			s.collector.Unlock()
+			time.Sleep(statsInterval)
+			s.collector.Lock()
+			continue
+		}
+		s.collector.Unlock()
+		element := s.collector.Flowentries.Remove(s.collector.Flowentries.Front())
+		//Now we can proceed lock free flowcache is not shared
+		_, err := s.FlowCache.Get(element.(*collectorentry).L4FlowHash)
+		if err != nil {
+			//this is new flow add it to our rpc payload
+			rpcpayload.NumFlows = rpcpayload.NumFlows + 1
+			rpcpayload.Flows = append(rpcpayload.Flows, *element.(*collectorentry).entry)
+		}
+		if time.Since(starttime) > statsInterval {
+			//Send out everything we have in the payload
+			request.Payload = rpcpayload
+			err = s.Rpchdl.RemoteCall(statsContextID, "StatsServer.GetStats", &request, &response)
+
+			starttime = time.Now()
+		}
+
+	}
+}
+func (s *Server) connectStatsClient(statsClient *StatsClient) error {
+
+	statschannel := os.Getenv(envStatsChannelPath)
+	err := statsClient.Rpchdl.NewRPCClient(statsContextID, statschannel)
+	if err != nil {
+		log.WithFields(log.Fields{"package": "remote_enforcer", "error": err}).Error("Stats RPC client cannot connect")
+	}
+	_, err = statsClient.Rpchdl.GetRPCClient(statsContextID)
+
+	go statsClient.SendStats()
```

Comment:
Even if you get errors from  statsClient.Rpchdl.NewRPCClient or statsClient.Rpchdl.GetRPCClient you still want to launch the goroutine ?